Class level __getattr__ causes error: __getattr__ is not valid at the module level outside a stub file

After updating to mypy 0.530, I am getting an error on a `__getattr__()` method defined on a class:

```
$ mypy --version
mypy 0.530
$ python3 --version
Python 3.6.3
[~/cretonne/lib/cretonne/meta]
$ cat mypy.ini 
[mypy]
disallow_untyped_defs = True
warn_unused_ignores = True
warn_return_any = True
[~/cretonne/lib/cretonne/meta]
$ mypy cdsl/formats.py
cdsl/formats.py:143: error: __getattr__ is not valid at the module level outside a stub file
```

I don't have any module level `__getattr__()` functions.

This didn't reproduce with a smaller test case, but you can check out the [Cretonne repo](https://github.com/stoklund/cretonne) to reproduce.

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -3326,7 +3326,7 @@ class Scope:
         top = self.top_function()
         assert top, "This method must be called from inside a function"
         index = self.stack.index(top)
-        assert index
+        assert index, "Scope stack must always start with a module"
         enclosing = self.stack[index - 1]
         if isinstance(enclosing, TypeInfo):
             return enclosing

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -259,15 +259,14 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
 
     def handle_cannot_determine_type(self, name: str, context: Context) -> None:
         node = self.scope.top_function()
-        if (self.pass_num < LAST_PASS and node is not None
-                and isinstance(node, (FuncDef, LambdaExpr))):
+        if self.pass_num < LAST_PASS and isinstance(node, (FuncDef, LambdaExpr)):
             # Don't report an error yet. Just defer.
             if self.errors.type_name:
                 type_name = self.errors.type_name[-1]
             else:
                 type_name = None
             # Shouldn't we freeze the entire scope?
-            active_class = self.scope.active_class()
+            active_class = self.scope.enclosing_class()
             self.deferred_nodes.append(DeferredNode(node, type_name, active_class))
             # Set a marker so that we won't infer additional types in this
             # function. Any inferred types could be bogus, because there's at
@@ -3321,6 +3320,15 @@ class Scope:
     def active_class(self) -> Optional[TypeInfo]:
         if isinstance(self.stack[-1], TypeInfo):
             return self.stack[-1]
+        return None
+
+    def enclosing_class(self) -> Optional[TypeInfo]:
+        top = self.top_function()
+        assert top, "This method must be called from inside a function"
+        index = self.stack.index(top)
+        for e in reversed(self.stack[:index]):
+            if isinstance(e, TypeInfo):
+                return e
         return None
 
     def active_self_type(self) -> Optional[Union[Instance, TupleType]]:

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -3326,9 +3326,10 @@ class Scope:
         top = self.top_function()
         assert top, "This method must be called from inside a function"
         index = self.stack.index(top)
-        for e in reversed(self.stack[:index]):
-            if isinstance(e, TypeInfo):
-                return e
+        assert index
+        enclosing = self.stack[index - 1]
+        if isinstance(enclosing, TypeInfo):
+            return enclosing
         return None
 
     def active_self_type(self) -> Optional[Union[Instance, TupleType]]:

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -266,8 +266,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
             else:
                 type_name = None
             # Shouldn't we freeze the entire scope?
-            active_class = self.scope.enclosing_class()
-            self.deferred_nodes.append(DeferredNode(node, type_name, active_class))
+            enclosing_class = self.scope.enclosing_class()
+            self.deferred_nodes.append(DeferredNode(node, type_name, enclosing_class))
             # Set a marker so that we won't infer additional types in this
             # function. Any inferred types could be bogus, because there's at
             # least one type that we don't know.

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -3995,3 +3995,29 @@ class E(metaclass=t.M): pass
 class F(six.with_metaclass(t.M)): pass
 @six.add_metaclass(t.M)
 class G: pass
+
+[case testCorrectEnclosingClassPushedInDeferred]
+class C:
+    def __getattr__(self, attr: str) -> int:
+        x: F
+        return x.f
+
+class F:
+    def __init__(self, f: int) -> None:
+        self.f = f
+[out]
+
+[case testCorrectEnclosingClassPushedInDeferred2]
+from typing import TypeVar
+T = TypeVar('T', bound=C)
+class C:
+    def m(self: T) -> T:
+        class Inner:
+            x: F
+            f = x.f
+        return self
+
+class F:
+    def __init__(self, f: int) -> None:
+        self.f = f
+[out]

--- a/test-data/unit/check-classes.test
+++ b/test-data/unit/check-classes.test
@@ -4022,6 +4022,15 @@ class F:
         self.f = f
 [out]
 
+[case testCorrectEnclosingClassPushedInDeferred3]
+class A:
+    def f(self) -> None:
+        def g(x: int) -> int:
+            return y
+
+y = int()
+[out]
+
 [case testMetaclassMemberAccessViaType]
 from typing import Type
 class M(type):

